### PR TITLE
Add makegocons command

### PR DESCRIPTION
**Pull request by chenglch**
This patch enable goconserver service and manage the node sessions
with a new `makegocons` command.

Implement: #4215

### CODE REVIEW

File: xCAT-server/lib/xcat/plugins/goconserver.pm [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+    $Getopt::Long::ignorecase = 0;
+    my $delmode;
+    GetOptions('d|delete' => \$delmode,
+    );
+
+    my $svboot = 0;
+    if (exists($req->{svboot})) {
+        $svboot = 1;
+    }
+    my @hostinfo = xCAT::NetworkUtils->determinehostname();
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/goconserver.pm]: The hostinfo is got here, why another one in line 182?

File: xCAT-server/lib/xcat/plugins/goconserver.pm [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+                }
+                $cmd = $locerror.$env.$cmd;
+            }
+            $data->{$k}->{driver} = "cmd";
+            $data->{$k}->{params}->{cmd} = $cmd;
+            $data->{$k}->{name} = $k;
+        }
+        if (defined($v->{consoleondemand})) {
+            # consoleondemand attribute for node can be "1", "yes", "0" and "no"
+            if ((($v->{consoleondemand} eq "1") || lc($v->{consoleondemand}) eq "yes") && !$siteondemand) {
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/goconserver.pm]: If node:consoleondemand equal 1, the $ondemain=1 whether siteondemand is set or not, right?
If node:consoleondemand equal 0,  the $ondemain=0 whether siteondemand is set or not, right?
If node:consoleondemand not set, if $siteondemand is set, the $ondemand shall 1, right?

File: xCAT-server/lib/xcat/plugins/goconserver.pm [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+        $openbmc_table->close();
+        foreach my $node (@openbmc_nodes) {
+            if (defined($openbmc_hash->{$node}->[0])) {
+                if (!$openbmc_hash->{$node}->[0]->{'bmc'}) {
+                    xCAT::SvrUtils::sendmsg("Error: Unable to get attribute bmc", $::callback, $node);
+                    delete $data->{$node};
+                    next;
+                }
+                $data->{$node}->{params}->{host} = $openbmc_hash->{$node}->[0]->{'bmc'};
+                if ($openbmc_hash->{$node}->[0]->{'username'}) {
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/goconserver.pm]: From my point of view, I think the admin use username/password either set in openbmc table or passwd table. It is invalid to use part of OpenBMC and passwd.

File: xCAT-server/lib/xcat/plugins/goconserver.pm [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+        {
+            foreach my $nodeent (keys %$ent)
+            {
+                push @tmpcons_nodes, $ent->{$nodeent}->[0];
+            }
+        }
+        @cons_nodes = @tmpcons_nodes
+
+    } else {
+        @cons_nodes = $hmtab->getAllNodeAttribs([ 'cons', 'serialport', 'mgt', 'conserver', 'consoleondemand' ]);
```
[the remark this answers is not in the document]
Comment (chenglch) [reply xCAT-server/lib/xcat/plugins/goconserver.pm]: Not sure, seems node is not needed. See https://github.com/xcat2/xcat-core/blob/master/xCAT-server/lib/xcat/plugins/conserver.pm#L368

File: xCAT-server/lib/xcat/plugins/goconserver.pm (hunk 2 above)
Comment (chenglch) [reply xCAT-server/lib/xcat/plugins/goconserver.pm]: Changed based on https://github.com/xcat2/xcat-core/blob/master/xCAT-server/lib/xcat/plugins/conserver.pm#L590, I will change my $ondemaind = \0; --> my $ondemaind = \1;

File: xCAT-server/lib/xcat/plugins/goconserver.pm (hunk 3 above)
Comment (chenglch) [reply xCAT-server/lib/xcat/plugins/goconserver.pm]: This is the code logic from https://github.com/xcat2/xcat-core/blob/master/xCAT-server/lib/xcat/plugins/openbmc.pm$#L1325

File: xCAT-server/lib/perl/xCAT/Goconserver.pm [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    }
+    my $response = http_request("DELETE", $url, $data);
+    if (!defined($response)) {
+        $rsp->{data}->[0] = "Failed to send delete request.";
+        xCAT::MsgUtils->message("E", $rsp, $callback);
+        return 1;
+    } elsif ($delmode) {
+        while (my ($k, $v) = each %{$response}) {
+            if ($v ne "Deleted") {
+                $rsp->{data}->[0] = "$k: Failed to delete delete entry in goconserver: $v";
```
[the remark this answers is not in the document]
Comment (chenglch) [reply xCAT-server/lib/perl/xCAT/Goconserver.pm]: Reply point 2 , if is false to return 1 here as this is in foreach loop we should reply message for every node.

File: xCAT-server/lib/perl/xCAT/Goconserver.pm [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                xCAT::MsgUtils->message("I", $rsp, $callback);
+            }
+        }
+    }
+    return 0;
+}
+
+sub create_nodes {
+    my ($api_url, $node_map, $callback) = @_;
+    my $url = "$api_url/bulk/nodes";
```
[the remark this answers is not in the document]
Comment (chenglch) [reply xCAT-server/lib/perl/xCAT/Goconserver.pm]: Can't the url is different, there are more rest urls that may be added here.

File: xCAT-server/lib/xcat/plugins/goconserver.pm [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+}
+
+sub get_cons_map {
+    my ($req, $iphashref) = @_;
+    my %cons_map;
+    my %iphash = %{$iphashref};
+    my $hmtab = xCAT::Table->new('nodehm');
+    my @cons_nodes;
+
+    if (($req->{node} and @{$req->{node}} > 0) or $req->{noderange}->[0]) {
```
[the remark this answers is not in the document]
Comment (chenglch) [reply xCAT-server/lib/xcat/plugins/goconserver.pm]: This logic has been exist for a very long time, see https://github.com/xcat2/xcat-core/blob/master/xCAT-server/lib/xcat/plugins/conserver.pm#L353 and https://github.com/xcat2/xcat-core/blob/master/xCAT-server/lib/xcat/plugins/conserver.pm#L331, I just use $req->{node} directly.

File: xCAT-server/lib/perl/xCAT/Goconserver.pm (hunk 5 above)
[the remark this answers is not in the document]
Comment (zet809) [reply xCAT-server/lib/perl/xCAT/Goconserver.pm]: I see, but you shall set $rc as 1, and return $rc in the end, right?

File: xCAT-server/lib/perl/xCAT/Goconserver.pm [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    my $response = http_request("POST", $url, $data);
+    if (!defined($response)) {
+        $rsp->{data}->[0] = "Failed to send create request.";
+        xCAT::MsgUtils->message("E", $rsp, $callback);
+        return 1;
+    } elsif ($response) {
+        while (my ($k, $v) = each %{$response}) {
+            if ($v ne "Created") {
+                $rsp->{data}->[0] = "$k: Failed to create console entry in goconserver: $v";
+                xCAT::MsgUtils->message("E", $rsp, $::callback);
```
Comment (zet809) [inline xCAT-server/lib/perl/xCAT/Goconserver.pm]: Shall have the similar logic with line 68

File: xCAT-server/lib/xcat/plugins/goconserver.pm [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+        if (lc($site_entry) eq "yes") {
+            $siteondemand = 1;
+        }
+        elsif (lc($site_entry) ne "no") {
+            # consoleondemand attribute is set, but it is not "yes" or "no"
+            xCAT::SvrUtils::sendmsg([ 1, "Unexpected value $site_entry for consoleondemand attribute in site table" ], $::callback);
+        }
+    }
+    my (@nodes);
+    my $data = gen_request_data(\%cons_map, $siteondemand);
```
Comment (zet809) [inline xCAT-server/lib/xcat/plugins/goconserver.pm]: Shall we check the result of $data and %cons_map?

**Comment by zet809** [comment]
Looks good to me, I agree to merge. Thx!

**Pull request merged**